airbyte-cdk/python/airbyte_cdk/sources/declarative/schema/dynamic_schema_loader.py

```diff
@@ -0,0 +1,112 @@
+#
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+#
+
+
+from dataclasses import InitVar, dataclass, field
+from typing import TYPE_CHECKING, Any, List, Mapping, Optional, Union
+
+import dpath
+from airbyte_cdk.sources.declarative.decoders.decoder import Decoder
+from airbyte_cdk.sources.declarative.decoders.json_decoder import JsonDecoder
+from airbyte_cdk.sources.declarative.interpolation.interpolated_string import InterpolatedString
+from airbyte_cdk.sources.declarative.requesters.requester import Requester
+from airbyte_cdk.sources.declarative.schema.schema_loader import SchemaLoader
+from airbyte_cdk.sources.types import Config
+
+if TYPE_CHECKING:
+    from airbyte_cdk.sources.declarative.declarative_stream import DeclarativeStream
+
+
+@dataclass(frozen=True)
+class TypesPair:
+    target_type: str
+    current_type: Union[InterpolatedString, str]
+
+
+@dataclass
+class DynamicSchemaLoader(SchemaLoader):
+    stream: "DeclarativeStream"
+    config: Config
+    schema_pointer: List[Union[InterpolatedString, str]]
+    key_pointer: List[Union[InterpolatedString, str]]
+    parameters: InitVar[Mapping[str, Any]]
+    types_map: List[TypesPair]
+    type_pointer: Optional[List[Union[InterpolatedString, str]]] = None
+    decoder: Decoder = field(default_factory=lambda: JsonDecoder(parameters={}))
+
+    def __post_init__(self, parameters: Mapping[str, Any]) -> None:
+        self.schema_pointer = self._update_pointer(pointer=self.schema_pointer, parameters=parameters)
+        self.key_pointer = self._update_pointer(pointer=self.key_pointer, parameters=parameters)
+        self.type_pointer = self._update_pointer(pointer=self.type_pointer, parameters=parameters) if self.type_pointer else None
+
+    @staticmethod
+    def _update_pointer(
+        pointer: Optional[List[Union[InterpolatedString, str]]], parameters
+    ) -> Optional[List[Union[InterpolatedString, str]]]:
+        _pointer = [InterpolatedString.create(path, parameters=parameters) for path in pointer]
+        for path_index in range(len(pointer)):
+            if isinstance(pointer[path_index], str):
+                _pointer[path_index] = InterpolatedString.create(pointer[path_index], parameters=parameters)
+
+        return _pointer
+
+    def _replace_type_if_not_valid(self, field_type):
+        for types_pair in self.types_map:
+            if field_type == types_pair.current_type:
+                return types_pair.target_type
+        return field_type
+
+    def get_json_schema(self) -> Mapping[str, Any]:
+        response_json = next(self.stream.read_only_records(), None)
```

Comment:
Regarding the schema_pointer, I agree. I'm trying to ensure we have the flexibility to point to the schema while retaining metadata from fields at higher levels of the record. But I'm ok just leave a extract records 

As for `key_pointer/type_pointer`, I'm not entirely sure I understand your concern. We don’t need the full data from the record—just the field name and type. I believe it’s more intuitive to point directly to the field without removing fields or performing extensive record transformations. This is especially true given that we’re limiting how `FieldPointer` is implemented, as we need the key to be interpolated.

Reply:
The `schema_pointer` must be included in the dynamic schema loader configuration. This is necessary because, in cases where schemas are provided only as a list, we need the metadata (e.g. id of schema) of each schema to filter the desired schema record from others.

Example response:
```
[
  { "schema_id": 1, "fields": [] },
  { "schema_id": 2, "fields": [] }
]
```

Reply:
The schema_pointer logic seems to be duplication from `DpathExtractor`. Why can't we set up the extractor on the stream level and read from the extracted records? Something like:
```
for property_definition in self.retriever.read_records({}):
  <extract from type_pointer>
```

> This is necessary because, in cases where schemas are provided only as a list, we need the metadata (e.g. id of schema) of each schema to filter the desired schema record from others.

Can't this be a filter on the stream as well?

Reply:
The extractor determines where records will be located and trims all data from levels above. For example, if we have a dataset like {name: 'test', schema: [{field: ''}]} and set the path to ['schema'], we will only retrieve the schema fields. This works if we receive only one schema in the response. However, if the response includes a list of schemas, we need to filter them by name or another field. These fields are typically located at the same level as the schema fields list. Therefore, we first need to extract all schema objects, filter them by the desired field, and then extract schema data along with their fields and types.